### ISSUE TITLE
Fails to resolve when same requirement specific multiple times

### ISSUE THREAD

**Issue by chiefnoah** [pantsbuild/pex#1550]
Error in pants/pex when trying to package multiple modules that depend on each other, with shared external dependencies.

```
AssertionError: More than one direct requirement is satisfied by PyJWT 1.7.1:
0. PyJWT
1. PyJWT==1.7.1
This should never happen since Pip fails when more than one requirement for a given project name key is supplied and applies for a given target interpreter environment.
```
constraints.txt
```
PyJWT>=1.7.1,<2.0
```
requirements.txt #1
```
PyJWT
```

requirements.txt #2
```
PyJWT==1.7.1
```

To clarify, none of these requirements constraints are mutually exclusive and `pip` **does not** error when multiple of the same requirement are specified. This could potentially be an issue with how pants is passing in the requirements.

### Full traceback

```
stderr:
Traceback (most recent call last):
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 476, in execute
    exit_value = self._wrap_coverage(self._wrap_profiling, self._execute)
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 401, in _wrap_coverage
    return runner(*args)
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 432, in _wrap_profiling
    return runner(*args)
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 532, in _execute
    return self.execute_entry(self._pex_info.entry_point)
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 668, in execute_entry
    return self.execute_pkg_resources(entry_point)
  File "/home/noah/.cache/pants/named_caches/pex_root/unzipped_pexes/01a8d22400b26b1d132a9849fb6e28962f1ecf9e/.bootstrap/pex/pex.py", line 700, in execute_pkg_resources
    return runner()
  File "/home/noah/.cache/pants/named_caches/pex_root/installed_wheels/7fea57708be4117c2d962e01abd1585a8846dc06/pex-2.1.54-py2.py3-none-any.whl/pex/bin/pex.py", line 666, in main
    do_main(
  File "/home/noah/.cache/pants/named_caches/pex_root/installed_wheels/7fea57708be4117c2d962e01abd1585a8846dc06/pex-2.1.54-py2.py3-none-any.whl/pex/bin/pex.py", line 687, in do_main
    pex_builder = build_pex(
  File "/home/noah/.cache/pants/named_caches/pex_root/installed_wheels/7fea57708be4117c2d962e01abd1585a8846dc06/pex-2.1.54-py2.py3-none-any.whl/pex/bin/pex.py", line 578, in build_pex
    result = resolve(
  File "/home/noah/.cache/pants/named_caches/pex_root/installed_wheels/7fea57708be4117c2d962e01abd1585a8846dc06/pex-2.1.54-py2.py3-none-any.whl/pex/resolver.py", line 943, in resolve
    build_and_install_request.install_distributions(
  File "/home/noah/.cache/pants/named_caches/pex_root/installed_wheels/7fea57708be4117c2d962e01abd1585a8846dc06/pex-2.1.54-py2.py3-none-any.whl/pex/resolver.py", line 707, in install_distributions
    raise AssertionError(
AssertionError: More than one direct requirement is satisfied by PyJWT 1.7.1:
0. PyJWT
1. PyJWT==1.7.1
This should never happen since Pip fails when more than one requirement for a given project name key is supplied and applies for a given target interpreter environment.
```

**Comment by jsirois** [comment]
@chiefnoah what version of Pants are you using?

**Comment by chiefnoah** [comment]
```
[GLOBAL]
pants_version = "2.9.0.dev2"
```

We're currently using `2.2.0`, but I'm attempting to bump the version again (because that's very old and deprecated), but any version newer than that appears to have this problem.

**Comment by jsirois** [comment]
Ok, thanks @chiefnoah. Could you share the PEX-INFO? If you're seeing this via `./pants run ...` you'd want to `./pants package ...` and then `unzip -qc dist/my.pex PEX-INFO`. If you get this error from some command other than `./pants run` can you provide a bit more detail? Perhaps the full logs from `./pants -ldebug --no-local-cache ...`.

**Comment by chiefnoah** [comment]
I'm seeing this when trying to run `./pants package ::` to package all of our modules, and also when I attempt to package the specific module that's failing here. Adding `-ldebug --no-local-cache` doesn't add any additional output. 

The actual command appears to be:

```
  ProcessExecutionFailure: Process 'Building redacted.pex with 10 requirements: Flask, PyJWT, PyJWT==1.7.1, cachetools, cryptography, fire>=0.1.3, protobuf, requests==2.25.1, tabulate==0.8.7, typing_extensions>=3.7.4.3' failed with exit code 1.

...
<above stacktrace>
```

I can attempt to package another module and pull the PEX-INFO out from there, but I'm not sure how much that will help.

**Comment by jsirois** [comment]
Thanks, that was enough. This is definitely a Pants bug passing the same requirement 2x.   I'll take a look over there and see if I can't see what conditions lead to that.

**Comment by jsirois** [comment]
Alright, Pants makes no attempt to dedup project names, just whole requirements; so `PyJWT` and `PyJWT==1.7.1` are treated as distinct:
https://github.com/pantsbuild/pants/blob/0b1805739af59f76b1eee86d263dc4c25fe4dcef/src/python/pants/backend/python/util_rules/pex.py#L130-L141

A person sitting down and using the Pex CLI would ~never do this. That said, it appears you're correct re `This should never happen since Pip fails when more than one requirement for a given project name key is supplied and applies for a given target interpreter environment.` and that's definitely False.

To be clear, what's happening here is at PEX file boot time, the Pex runtime is trying to determine which distributions contained in the PEX file to activate and add to the sys.path. In so doing, its complaining about the same requirement being recorded 2x in its PEX-INFO. Let me dig in the Pex logic a bit here and see what can be done. This is not simple since a common case of multiple requirements is when you use environment markers to express "Choose this version for this environment and that version for that one.". In those cases you cannot collapse dup requirements but there is also no problem at runtime since only one of the two environments is active / present.

**Comment by jsirois** [comment]
Aha, the `This should never happen ...` bit is, at least in part, true:
```
$ pex -V
2.1.59
$ pex --resolver-version pip-legacy-resolver PyJWT PyJWT==1.7.1 -o PyJWT.pex --include-tools
ERROR: Double requirement given: PyJWT==1.7.1 (already in PyJWT, name='PyJWT')
pid 25018 -> /home/jsirois/.pex/venvs/52ef3fa0fec32d58685d2164c25fd2d092f23324/9148e449944f6676b304f8f06462e05782fdb928/pex --disable-pip-version-check --no-python-version-warning --exists-action a --use-deprecated legacy-resolver --isolated -q --cache-dir /home/jsirois/.pex download --dest /tmp/tmp5mn2ez7g/home.jsirois..venv.pex.bin.python3 PyJWT PyJWT==1.7.1 --index-url https://pypi.org/simple --retries 5 --timeout 15 exited with 1 and STDERR:
None
```

It's only using the new resolver, which Pants now does unconditionally, that it's false:
```
$ pex --resolver-version pip-2020-resolver PyJWT PyJWT==1.7.1 -o PyJWT.pex --include-tools
Traceback (most recent call last):
  File "/home/jsirois/.venv/pex/bin/pex", line 8, in <module>
    sys.exit(main())
  File "/home/jsirois/.venv/pex/lib/python3.10/site-packages/pex/bin/pex.py", line 683, in main
    do_main(
  File "/home/jsirois/.venv/pex/lib/python3.10/site-packages/pex/bin/pex.py", line 704, in do_main
    pex_builder = build_pex(
  File "/home/jsirois/.venv/pex/lib/python3.10/site-packages/pex/bin/pex.py", line 593, in build_pex
    result = resolve(
  File "/home/jsirois/.venv/pex/lib/python3.10/site-packages/pex/resolver.py", line 958, in resolve
    build_and_install_request.install_distributions(
  File "/home/jsirois/.venv/pex/lib/python3.10/site-packages/pex/resolver.py", line 722, in install_distributions
    raise AssertionError(
AssertionError: More than one direct requirement is satisfied by PyJWT 1.7.1:
0. PyJWT
1. PyJWT==1.7.1
This should never happen since Pip fails when more than one requirement for a given project name key is supplied and applies for a given target interpreter environment.
```

**Comment by chiefnoah** [comment]
Interesting. For some additional context, what's effectively happening here is we have module A and module B. Module A depends on module B, and both depend on 3rd-party dependency C. Both module A and B exist in the same repo as independent source_roots and packaged separately (one as a `.pex`, another as `.whl`). It seems Pants tries to union all requirements from both modules at package-time and resolve the environment for the `.pex` (which makes sense), but it's not de-duplicating the common dependency. The existence of a `constraints.txt` should disambiguate the version it ultimately resolves and passes to `pex`. 

Should I create an issue on the pants repo?

**Comment by jsirois** [comment]
> Should I create an issue on the pants repo?

I don't think so. You really can't dedup unless you are a resolver. Neither Pants nor Pex are resolvers. Pants delegates to Pex and Pex to Pip. Now Pip has enough context to dedup; i.e.: Its supposed to have enough logic to understand that `PyJWT` and `PyJWT==1.7.1` always dedups to `PyJWT==1.7.1` but that `PyJWT==1.7.1; python_version <= "3.7"` and `PyJWT>1.7.1; python_version > "3.7"` de-dups to one or the other depending on the interpreter being used to perform the resolve, etc.

**Comment by jsirois** [comment]
I'll get this fixed up in Pex 2.1.60 and upgrade Pants to that for the 2.9.x series (It's on 2.9.0rc0 presently).

**Comment by jsirois** [comment]
@chiefnoah this is now released in Pex 2.1.60, is in Pants main and is being cherry-picked to 2.9.x here: https://github.com/pantsbuild/pants/pull/13949. This will be released in Pants 2.9.0rc2.
